src/Psy/TabCompletion/AutoCompleter.php

```diff
@@ -82,18 +87,37 @@ public function processCallback($input, $index, $info = array())
     }
 
     /**
+     * @param string $line  Readline current line
+     * @param int    $start The start of the readline input
+     * @param int    $end   The end of the readline input
+     *
+     * @return bool true if the characters before $index are only whitespace
+     *
+     * @internal param int $index Current word index
+     */
+    private function shouldInsertOnlyTab($line, $start, $end)
+    {
+        $precedingInput = substr($line, 0, $end);
+
+        $trimmedInput = trim($precedingInput);
+
+        return empty($trimmedInput);
+    }
+
+    /**
      * The readline_completion_function callback handler.
      *
      * @see processCallback
      *
-     * @param $input
-     * @param $index
+     * @param string $input
+     * @param int    $start
+     * @param int    $end
      *
      * @return array
      */
-    public function callback($input, $index)
+    public function callback($input, $start, $end)
```

Comment:
do all the possible combinations of readline and libedit and php versions all provide `$end` for this callback? documentation is sadly lacking here :(

if not, should we fall back to the end in `readline_info` if it's available?

Reply:
I looked here: https://github.com/php/php-src/blob/master/ext/readline/readline.c#L490

And I saw that `params` always contains `int start` and `int end`.

For editline / libedit, editline/readline.h is included. This has `typedef char	**rl_completion_func_t(const char *, int, int);` and `extern rl_completion_func_t *rl_attempted_completion_function;`

in editline's readline.c I see this:
```
/*
 * complete word at current point
 */
/* ARGSUSED */
int
rl_complete(int ignore __attribute__((__unused__)), int invoking_key)
{
	static ct_buffer_t wbreak_conv, sprefix_conv;
	char *breakchars;

	if (h == NULL || e == NULL)
		rl_initialize();

	if (rl_inhibit_completion) {
		char arr[2];
		arr[0] = (char)invoking_key;
		arr[1] = '\0';
		el_insertstr(e, arr);
		return CC_REFRESH;
	}

	if (rl_completion_word_break_hook != NULL)
		breakchars = (*rl_completion_word_break_hook)();
	else
		breakchars = rl_basic_word_break_characters;

	_rl_update_pos();

	/* Just look at how many global variables modify this operation! */
	return fn_complete(e,
	    (rl_compentry_func_t *)rl_completion_entry_function,
	    rl_attempted_completion_function,
	    ct_decode_string(rl_basic_word_break_characters, &wbreak_conv),
	    ct_decode_string(breakchars, &sprefix_conv),
	    _rl_completion_append_character_function,
	    (size_t)rl_completion_query_items,
	    &rl_completion_type, &rl_attempted_completion_over,
	    &rl_point, &rl_end);


}
```

Where `rl_point` and `rl_end` are updated as follows:
```
static void
_rl_update_pos(void)
{
	const LineInfo *li = el_line(e);

	rl_point = (int)(li->cursor - li->buffer);
	rl_end = (int)(li->lastchar - li->buffer);
}
```

Which leads me to believe that it does the same thing.

Even if that is in question, from the `readline.c` file in php-src, we can see that rl_end (the value passed to the callback as end) and the `end` key output by `readline_info` are the same:

https://github.com/php/php-src/blob/master/ext/readline/readline.c#L257

Reply:
However, having dived into it this far, I cannot help but wonder how this works - I had interpreted `end` differently than what the documentation suggests it does, and I saw it work in that way as well. The variable I actually need is `rl_point`, which is always in `readline_info` but not passed as a callback argument.